php-checks/src/main/java/org/sonar/php/checks/MultilineBlocksCurlyBracesCheck.java

```diff
@@ -26,104 +26,166 @@
 
-  private void checkStatement(StatementTree firstInnerStatement, int nestingStatementNum, List<StatementTree> statements) {
-    if (firstInnerStatement.is(Kind.BLOCK)) {
-      return;
-    }
+    int lastStatementLine;
 
-    int firstIndent = column(firstInnerStatement);
-    List<StatementTree> statementsWhichShouldBeNested = new ArrayList<>();
+    int lastStatementColumn;
 
-    for (int i = nestingStatementNum + 1; i < statements.size(); i++) {
-      StatementTree nextStatement = statements.get(i);
+    boolean firstStatementOneLiner;
 
-      if (!nextStatement.is(Kind.INLINE_HTML)) {
+    StatementBlock(StatementTree firstStatement, int parentMarginColumn, int parentLastLine) {
+      this.firstStatement = firstStatement;
+      lastStatementLine = line(firstStatement);
+      this.firstStatementOneLiner = lastStatementLine == parentLastLine;
+      lastStatementColumn = column(firstStatement);
+      this.parentMarginColumn = parentMarginColumn;
+    }
 
-        if (column(nextStatement) == firstIndent) {
-          statementsWhichShouldBeNested.add(nextStatement);
+    int size() {
+      return 1 + otherStatement.size();
+    }
 
-        } else {
-          break;
-        }
+    /**
+     * @return false if the statement is not part of the group
+     */
+    boolean add(StatementTree statement) {
+      if (statement.is(Kind.INLINE_HTML)) {
+        // ignore and continue
+        return true;
+      }
+      int line = line(statement);
+      int column = column(statement);
+      if (column < parentMarginColumn || !isLineCloseToPreviousStatement(line)) {
+        return false;
+      } else if (column == parentMarginColumn &&
+        (firstStatementOneLiner || column != lastStatementColumn || !isOneLineAfterPreviousStatement(line))) {
+        return false;
+      } else {
+        otherStatement.add(statement);
+        lastStatementLine = line;
+        lastStatementColumn = column;
+        return true;
       }
     }
 
-    if (!statementsWhichShouldBeNested.isEmpty()) {
-      int firstInnerStatementLine = ((PHPTree) firstInnerStatement).getLine();
-      int lastShouldBeNestedLine = ((PHPTree) statementsWhichShouldBeNested.get(statementsWhichShouldBeNested.size() - 1)).getLine();
-      int blockSize = lastShouldBeNestedLine - firstInnerStatementLine + 1;
-      String message = statements.get(nestingStatementNum).is(Kind.IF_STATEMENT) ? MESSAGE_IF : MESSAGE_LOOP;
-      context().newIssue(this, statementsWhichShouldBeNested.get(0), String.format(message, blockSize));
+    private boolean isLineCloseToPreviousStatement(int line) {
+      return line < (lastStatementLine + MAX_STATEMENT_DISTANCE_IN_THE_SAME_BLOCK);
     }
-  }
 
+    private boolean isOneLineAfterPreviousStatement(int line) {
+      return line == (lastStatementLine + 1);
+    }
+
+  }
 
   private static int column(Tree tree) {
     return ((PHPTree) tree).getFirstToken().column();
   }
+
+  private static int line(Tree tree) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Move this method into "StatementBlock". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3398)